File: cloud_flight.py
import logging

import threading
import hid
import schedule

vendor_id = 0x0951
product_id = 0x16C4


def trigger_battery_level_response():
    with hid.Device(vendor_id, product_id) as device:
        device.write(
            bytes(
                [
                    0x21,
                    0xFF,
                    0x05,
                    0x00,
                    0x00,
                    0x00,
                    0x00,
                    0x00,
                    0x00,
                    0x00,
                    0x00,
                    0x00,
                    0x00,
                    0x00,
                    0x00,
                    0x00,
                    0x00,
                    0x00,
                    0x00,
                    0x00,
                ]
            )
        )


def battery_level(charge_state, value):
    logging.debug("Charge state: {} # value: {}".format(charge_state, value))
    # charging
    if charge_state == 0x10 or charge_state == 0x11:
        return 101

    if charge_state == 0xF:
        if value >= 130:
            return 100
        if value < 130 and value >= 120:
            return 95
        if value < 120 and value >= 100:
            return 90
        if value < 100 and value >= 70:
            return 85
        if value < 70 and value >= 50:
            return 80
        if value < 50 and value >= 20:
            return 75
        if value < 20 and value > 0:
            return 70

    if charge_state == 0xE:
        if value > 240:
            return 65
        if value < 240 and value >= 220:
            return 60
        if value < 220 and value >= 208:
            return 55
        if value < 208 and value >= 200:
            return 50
        if value < 200 and value >= 190:
            return 45
        if value < 190 and value >= 180:
            return 40
        if value < 179 and value >= 169:
            return 35
        if value < 169 and value >= 159:
            return 30
        if value < 159 and value >= 148:
            return 25
        if value < 148 and value >= 119:
            return 20
        if value < 119 and value >= 90:
            return 15
        if value < 90:
            return 10
    
    return 0


def volume_direction(value):
    if value == 0x01:
        return "+"
    elif value == 0x02:
        return "-"
    return None


class CloudFlightThread(threading.Thread):
    def __init__(self, on_power, on_mute, on_battery, on_volume):
        super(CloudFlightThread, self).__init__(
            target=self._collect,
            args=[on_power, on_mute, on_battery, on_volume],
            daemon=True,
        )
        trigger_battery_level_response()
        schedule.every(5).minutes.do(trigger_battery_level_response)

    def _collect(self, on_power, on_mute, on_battery, on_volume):
        while True:
            for hid_device in hid.enumerate(vendor_id, product_id):
                with hid.Device(path=hid_device["path"]) as device:
                    data = device.read(2048, 100)

                    # power on/off/muted
                    if len(data) == 2:
                        if data[0] == 0x64:
                            if data[1] == 0x1:
                                trigger_battery_level_response()
                                on_power(True)
                            elif data[1] == 0x3:
                                on_power(False)

                        if data[0] == 0x65:
                            on_mute(data[1] == 0x04)
                    # volume direction
                    elif len(data) == 5:
                        vol_dir = volume_direction(data[1])
                        if vol_dir is not None:
                            on_volume(vol_dir)
                    # battery status
                    elif len(data) == 20:
                        on_battery(battery_level(data[3], data[4]))
            schedule.run_pending()

File: src/cloud_flight.rs
use hidapi::{HidApi, HidDevice};
use log::{debug, info};
use std::cell::Cell;

const VENDOR_ID: u16 = 0x0951;
const PRODUCT_IDS: [u16; 2] = [0x1723, 0x16c4];

const BATTERY_TRIGGER_PACKET: [u8; 20] = {
    let mut buf = [0; 20];
    buf[0] = 0x21;
    buf[1] = 0xff;
    buf[2] = 0x05;
    buf
};

fn battery_percent(charge_state: u8, value: u8) -> u8 {
    match charge_state {
        0x0e => match value {
            0..=89 => 10,
            90..=119 => 15,
            120..=148 => 20,
            149..=159 => 25,
            160..=169 => 30,
            170..=179 => 35,
            180..=189 => 40,
            190..=199 => 45,
            200..=209 => 50,
            210..=219 => 55,
            220..=239 => 60,
            240..=255 => 65,
        },
        0x0f => match value {
            0..=19 => 70,
            20..=49 => 75,
            50..=69 => 80,
            70..=99 => 85,
            100..=119 => 90,
            120..=129 => 95,
            130..=255 => 100,
        },
        _ => 0,
    }
}

pub enum Event {
    Battery { value: u8 },
    BatteryCharging,
    VolumeUp,
    VolumeDown,
    Muted,
    Unmuted,
    PowerOff,
    PowerOn,
    Ignored,
}
pub struct CloudFlight {
    device: HidDevice,
    pub powered: Cell<bool>,
    pub muted: Cell<bool>,
    pub charging: Cell<bool>,
    pub battery: Cell<u8>,
}
impl CloudFlight {
    pub fn new() -> Self {
        let api = HidApi::new().unwrap();

        let device = PRODUCT_IDS
            .iter()
            .map(|pid| api.open(VENDOR_ID, *pid))
            .filter(|device| device.is_ok())
            .map(|device| device.unwrap())
            .last();

        if device.is_none() {
            panic!("Not found any compatible device");
        }

        CloudFlight {
            device: device.unwrap(),
            powered: Cell::new(true),
            muted: Cell::new(false),
            charging: Cell::new(false),
            battery: Cell::new(100),
        }
    }
    pub fn read(&self) -> Event {
        let mut buf = [0u8; 32];
        let bytes = self.device.read_timeout(&mut buf, 500).unwrap();
        debug!("Read: {}, {:02x?}", bytes, buf);
        match bytes {
            2 => {
                if buf[0] == 0x64 {
                    if buf[1] == 0x01 {
                        self.battery();
                        self.powered.set(true);
                        info!("Power on");
                        return Event::PowerOn;
                    } else if buf[1] == 0x03 {
                        self.powered.set(false);
                        info!("Power off");
                        return Event::PowerOff;
                    }
                }
                if buf[0] == 0x65 {
                    if buf[1] == 0x04 {
                        self.muted.set(true);
                        info!("Muted");
                        return Event::Muted;
                    } else {
                        self.muted.set(false);
                        info!("Unmuted");
                        return Event::Unmuted;
                    }
                }
                return Event::Ignored;
            }
            5 => {
                if buf[1] == 0x01 {
                    info!("Volume up");
                    return Event::VolumeUp;
                } else if buf[1] == 0x02 {
                    info!("Volume down");
                    return Event::VolumeDown;
                }
                return Event::Ignored;
            }
            20 => {
                if buf[3] == 0x10 || buf[3] == 0x11 {
                    info!("Battery charging");
                    self.charging.set(true);
                    if buf[4] >= 20 {
                        return Event::BatteryCharging;
                    }
                    return Event::Battery { value: 100 };
                }
                let b_percent = battery_percent(buf[3], buf[4]);
                info!("Battery {}", b_percent);
                self.charging.set(false);
                self.battery.set(b_percent);
                return Event::Battery { value: b_percent };
            }
            _ => return Event::Ignored,
        }
    }
    pub fn battery(&self) {
        self.device.write(&BATTERY_TRIGGER_PACKET).unwrap();
    }
}

unsafe impl Sync for CloudFlight {}
